scipy/optimize/_milp.py

```diff
@@ -0,0 +1,280 @@
+import numpy as np
+from scipy.sparse import csc_array
+from ._highs._highs_wrapper import _highs_wrapper
+from ._constraints import LinearConstraint, Bounds
+from ._optimize import OptimizeResult
+
+
+def _milp_iv(c, integrality, bounds, constraints, options):
+
+    c = np.atleast_1d(c).astype(np.double)
+    if c.ndim != 1:
+        message = "`c` must be a one-dimensional array."
+        raise ValueError(message)
+
+    if integrality is None:
+        integrality = 0
+    integrality = np.broadcast_to(integrality, c.shape).astype(np.uint8)
+
+    if bounds is None:
+        bounds = Bounds(0, np.inf)
+    elif not isinstance(bounds, Bounds):
+        message = "`bounds` must be an instance of `scipy.optimize.Bounds`."
+        try:
+            bounds = Bounds(*bounds)
+        except TypeError:
+            raise ValueError(message)
+    # broadcast in case lb/ub are scalars
+    lb = np.broadcast_to(bounds.lb, c.shape).astype(np.double)
+    ub = np.broadcast_to(bounds.ub, c.shape).astype(np.double)
+
+    if constraints is None:
+        constraints = LinearConstraint(np.empty((0, c.size)),
+                                       np.empty((0,)), np.empty((0,)))
+    elif not isinstance(constraints, LinearConstraint):
+        message = ("`constraints` must be an instance of "
+                   "`scipy.optimize.LinearConstraint`.")
+        try:
+            constraints = LinearConstraint(*constraints)
+        except TypeError:
+            raise ValueError(message)
+    A = csc_array(constraints.A)
+    b_l = np.atleast_1d(constraints.lb).astype(np.double)
+    b_u = np.atleast_1d(constraints.ub).astype(np.double)
+    if b_l.ndim > 1 or b_l.shape != b_u.shape:
+        message = "`b_l` and `b_u` must be one-dimensional of equal length."
+        raise ValueError(message)
+    if A.shape != (b_l.size, c.size):
+        message = "The shape of `A` must be (len(b_l), len(c))."
+        raise ValueError(message)
+    indptr, indices, data = A.indptr, A.indices, A.data.astype(np.double)
+
+    # relying on options input validation in _highs_wrapper
+    options_iv = {}
+    options_iv.update(options)
+
+    return c, integrality, lb, ub, indptr, indices, data, b_l, b_u, options_iv
+
+
+def milp(c, *, integrality=None, bounds=None, constraints=None, options={}):
```

Comment:
mutable `options={}` should be made immutable

Reply:
I remember that there is only one default options object regardless of how many times `milp` is called. Is that why? What is the concern? Are we modifying it internally and that could cause something unintended to happen?

Reply:
Done.